Feat/redemption queue

kent-3 opened the pull request.

kent-3 committed:
wip: initial redeem queue contract implementation

kent-3 committed:
fix: typo

kent-3 committed:
wip: generate schema and ts types for redeem-queue-proxy

kent-3 committed:
wip: redeem-queue second pass

kent-3 committed:
add: redeem-queue-proxy contract

kent-3 committed:
test: redeem-queue-proxy contract

chris-ricketts commented on contracts/proxy/redeem-queue/contract.rs:
```diff
@@ -54,42 +60,211 @@ fn vault_reserve_balance(deps: Deps, hub: &str, vault: &str) -> Result<u128, Err
+        );
+    }
+
+    let process_head_response = handle_process_head(deps.branch(), vault.clone())?;
+
+    let mut queue = RedemptionQueue::new(deps.storage, &vault);
+    let index = queue.enqueue(info.sender.as_ref(), coin.amount)?;
+
+    let mut response = Response::default()
+        .add_attribute("kind", "redeem")
+        .add_attribute("address", &info.sender)
+        .add_attribute("vault", &vault)
+        .add_attribute("amount", coin.amount)
+        .add_attribute("entry_index", index.to_string());
+
+    for msg in process_head_response.messages {
+        response = response.add_submessage(msg);
+    }
+
+    let vault_metadata: VaultMetadata = deps.querier.query_wasm_smart(
+        &hub,
+        &HubQueryMsg::VaultMetadata {
+            vault: vault.to_owned(),
+        },
+    )?;
+    let reserve_balance = vault_metadata.reserve_balance.u128();
+    let available_amount = Uint128::new(reserve_balance);
+
+    // NOTE: Using 1 instead of 0 to accommodate precision errors
+    if available_amount > Uint128::one() {
+        let (processed, _used_amount) = queue.process_head(available_amount)?;
+
+        if !processed.is_empty() {
+            for (address, amount) in processed {
+                let msg = redeem_on_behalf(&hub, &vault, &synthetic, address, amount.u128());
+                response = response.add_message(msg);
+            }
+        }
+    }
```
For the redeem handler it seems we are doing:
1. Process Queue Head
2. Enqueue Entry
3. Process Queue Head (in a slightly different way)

This could lead to TX failures with the error "insufficient reserve balance" as the same `available_balance` is used in both 'Process Queue Head' operations. I.e. if the response to the first `handle_process_head` call contains a `RedeemOnBehalf` message, the second call to `RedemptionQueue::process_head` may try to "double spend" the `VaultMetadata::reserve_balance`. 

If possible we should simplify to:
1. Enqueue Entry 
2. Process Queue Head

If for some reason 'Process Queue Head' is required to run both before and after 'Enqueue Entry', the `available_balance` for the second 'Process Queue Head' operation should be reduced by any used amount in the first operation.

kent-3 replied:
Good catch! I hadn't considered that fact that the vault reserve balance query would be same since the submessages haven't actually happened yet.

chris-ricketts commented on contracts/proxy/redeem-queue/contract.rs:
```diff
@@ -54,42 +60,211 @@ fn vault_reserve_balance(deps: Deps, hub: &str, vault: &str) -> Result<u128, Err
+    let mut queue = RedemptionQueue::new(deps.storage, &vault);
+    let cancelled = queue.cancel_user_entries(info.sender.as_ref())?;
+
+    if cancelled.is_empty() {
+        return Ok(Response::default()
+            .add_attribute("kind", "cancel_all")
+            .add_attribute("address", &info.sender)
+            .add_attribute("vault", &vault)
+            .add_attribute("count", "0"));
+    }
```
I think we should return an error in this case.

If someone issued this message expecting to cancel one or more entries, cancelling no entries would be unexpected and should be handled as an error.

kent-3 replied:
I could go either way here. I thought of it like "I want to cancel all my entries regardless of how many I have" but it's no problem to change it.

chris-ricketts commented on contracts/proxy/redeem-queue/contract.rs:
```diff
@@ -54,42 +60,211 @@ fn vault_reserve_balance(deps: Deps, hub: &str, vault: &str) -> Result<u128, Err
+    let synthetic = synthetic_for_vault(deps.as_ref(), &hub, &vault)?;
+
+    let mut queue = RedemptionQueue::new(deps.storage, &vault);
+    let available_amount = Uint128::new(reserve_balance);
```
We could reduce this to a single query:

```suggestion
    let vault_metadata: VaultMetadata = deps.querier.query_wasm_smart(
        &hub,
        &HubQueryMsg::VaultMetadata {
            vault: vault.to_owned(),
        },
    )?;
    let synthetic = vault_metadata.synthetic;
    let available_amount = vault_metadata.reserve_balance;
    
    let mut queue = RedemptionQueue::new(deps.storage, &vault);
```

chris-ricketts commented on contracts/proxy/redeem-queue/contract.rs:
```diff
@@ -54,42 +60,211 @@ fn vault_reserve_balance(deps: Deps, hub: &str, vault: &str) -> Result<u128, Err
+        }
+    }
+
+    Ok(response)
```
This response should also contain the `processed_entries` and `used_amount` attributes from the 'Process Queue Head' operation.

kent-3 replied:
Agreed. In my head they were being emitted as separate events, but not in reality :laughing:

chris-ricketts reviewed (commented):
Nice work @kent-3! 

The integration tests are especially thorough. Thanks for bearing with all the setup boilerplate - ideally at some point that burden in reduced as much as possible. 

I have just left a couple of comments, feel free to reply when you can.

kent-3 committed:
use single query for vault metadata

Co-authored-by: Chris Ricketts <6156768+chris-ricketts@users.noreply.github.com>

kent-3 committed:
PR comments

* comment #1

enqueue before processing head to avoid "double spend" condition

* comment #2

return Error when attempting to cancel_all if there are no entries

* add response attributes even if processed.is_empty()

chris-ricketts commented:
LGTM! Thanks @kent-3

I'll fix the formatting which is failing the CI check after the merge.
